overlord/assertstate/assertstate.go

```diff
@@ -538,6 +538,138 @@ func ValidationSetAssertionForMonitor(st *state.State, accountID, name string, s
+		}
+		if err != nil {
+			return nil, err
+		}
+		vs = a.(*asserts.ValidationSet)
+		return vs, nil
+	}
+
+	// try to get existing from the db. It will be the latest one if it was
+	// tracked already and thus refreshed via RefreshValidationSetAssertions.
+	// Otherwise, it may be a local assertion that was tracked in the past and
+	// then forgotten, in which case we need to refresh it explicitly.
+	db := cachedDB(st)
+	headers := map[string]string{
+		"series":     release.Series,
+		"account-id": accountID,
+		"name":       name,
+	}
+	if sequence > 0 {
+		headers["sequence"] = fmt.Sprintf("%d", sequence)
+	}
+
+	pool := asserts.NewPool(db, maxGroups)
+	atSeq := &asserts.AtSequence{
+		Type:        asserts.ValidationSetType,
+		SequenceKey: []string{release.Series, accountID, name},
+		Sequence:    sequence,
+		Revision:    asserts.RevisionNotKnown,
+		Pinned:      sequence > 0,
+	}
+
+	vs, err = getSpecificSequenceOrLatest(db, headers)
+
+	// found locally
+	if err == nil {
+		// check if we were tracking it already; if not, that
+		// means we found an old assertion (it was very likely tracked in the
+		// past) and we need to update it as it wasn't covered
+		// by RefreshValidationSetAssertions.
+		var tr ValidationSetTracking
+		trerr := GetValidationSet(st, accountID, name, &tr)
+		if trerr != nil && trerr != state.ErrNoState {
+			return nil, trerr
+		}
+		// not tracked, the assertion.
+		if trerr == state.ErrNoState {
+			// update with pool
+			atSeq.Sequence = vs.Sequence()
+			atSeq.Revision = vs.Revision()
+			if err := pool.AddSequenceToUpdate(atSeq, atSeq.Unique()); err != nil {
+				return nil, err
+			}
+		} else {
+			// was already tracked, add to validation sets and check
+			if err := valsets.Add(vs); err != nil {
+				return nil, err
+			}
+			if err := valsets.Conflict(); err != nil {
+				return nil, fmt.Errorf("cannot enforce validation set: %v", err)
+			}
+			if err := valsets.CheckInstalledSnaps(snaps); err != nil {
+				return nil, fmt.Errorf("cannot enforce validation set: %v", err)
+			}
+			return vs, nil
+		}
+	} else {
+		if !asserts.IsNotFound(err) {
+			return nil, err
+		}
+
+		// try to resolve with pool
+		if err := pool.AddUnresolvedSequence(atSeq, atSeq.Unique()); err != nil {
+			return nil, err
+		}
+	}
+
+	checkForConflicts := func(db *asserts.Database, bs asserts.Backstore) error {
+		tmpDb := db.WithStackedBackstore(bs)
+		// get the resolved validation set assert, add to validation sets and check
+		vs, err = getSpecificSequenceOrLatest(tmpDb, headers)
+		if err != nil {
+			return fmt.Errorf("internal error: cannot find validation set assertion: %v", err)
+		}
+		if err := valsets.Add(vs); err != nil {
+			return fmt.Errorf("internal error: cannot check validation sets conflicts: %v", err)
+		}
+		if err := valsets.Conflict(); err != nil {
+			return err
+		}
+		if err := valsets.CheckInstalledSnaps(snaps); err != nil {
+			return err
+		}
```

Comment:
and this should probably be a shared helper, also because I'm not sure it's we want to wrap the error in one case but not here?

Reply:
Thanks, good idea, done. As for wrapping I got rid of it, I think it's actually better to leave this to the caller and have access to actual error type/details if needed.